File: javascript/binary-tree-inorder-traversal.js
'use strict';

const binaryTreeInorderTraversal = (root) => {
    const result = [];
    const stack = [];
    let current = root;
    
    while (current || stack.length) {
        if (current) {
        stack.push(current);
        current = current.left;
        } else {
        current = stack.pop();
        result.push(current.val);
        current = current.right;
        }
    }
    
    return result;
};

const root = {
    val: 1,
    left: null,
    right: {
        val: 2,
        left: {
        val: 3,
        left: null,
        right: null,
        },
        right: null,
    },
};

console.log(binaryTreeInorderTraversal(root)); // [1, 3, 2]

File: go/btree/binary-tree-inorder-traversal.go
package btree

import "fmt"

func inOrderTraversal(node *Node) []int {
	var result []int
	var stack []*Node
	if node == nil {
		return result
	}
	for node != nil || len(stack) > 0 {
		for node != nil {
			stack = append(stack, node)
			node = node.Left
		}
		if len(stack) > 0 {
			node = stack[len(stack)-1]
			result = append(result, node.Val)
			stack = stack[:len(stack)-1]
			node = node.Right
		}

	}
	return result
}

func RunInOrderTraversal() {
	tree := Tree{
		Root: &Node{
			Val: 1,
			Left: &Node{
				Val: 2,
			},
			Right: &Node{
				Val: 3,
				Left: &Node{
					Val: 4,
				},
			},
		},
	}
	fmt.Println(inOrderTraversal(tree.Root))
}
